File: packages/modules/api/python/src/api/gemini.google.com/gemini.py
import httpx

from api.types import GeminiModel as Model
from api.types import GeminiContent

BASE_URL = "https://generativelanguage.googleapis.com"


class GeminiClient:
    def __init__(self, api_key: str) -> None:
        self._api_key = api_key
        self._client = httpx.AsyncClient()

    async def generate_content(
        self, contents: list[GeminiContent], model: Model = Model.GEMINI_2_0_FLASH, timeout_secs: int = 60
    ) -> dict:
        url = f"{BASE_URL}/v1beta/models/{model.value}:generateContent?key={self._api_key}"
        body = {"contents": [{"role": c.role.value, "parts": [{"text": p.text} for p in c.parts]} for c in contents]}
        response = await self._client.post(url, json=body, timeout=timeout_secs)
        return response.json()

File: packages/cli/jack/cli.kt/src/main/kotlin/io/github/hieudoanm/cli/commands/Gemini.kt
package io.github.hieudoanm.cli.commands

import com.github.ajalt.clikt.core.CliktCommand
import com.github.ajalt.clikt.core.subcommands
import com.github.ajalt.clikt.parameters.arguments.argument
import com.google.gson.GsonBuilder
import com.google.gson.JsonParser
import io.github.hieudoanm.cli.services.Requests

class GeminiCommand : CliktCommand(name = "gemini", help = "Interact with Google Gemini AI models") {
    init {
        subcommands(GeminiCode())
    }
    override fun run() = Unit
}

class GeminiCode : CliktCommand(name = "code", help = "Gemini-powered AI coding assistant") {
    private val prompt by argument()

    override fun run() {
        val apiKey = System.getenv("GEMINI_API_KEY")
        if (apiKey.isNullOrEmpty()) {
            echo("GEMINI_API_KEY environment variable is not set", err = true)
            return
        }

        val url = "https://generativelanguage.googleapis.com/v1/models/gemini-2.0-flash:generateContent?key=$apiKey"
        val requestBody = """{"contents":[{"parts":[{"text":${GsonBuilder().create().toJson(prompt)}}]}]}"""
        val result = Requests.post(url, requestBody, mapOf("Content-Type" to "application/json"))
        if (result.isFailure) {
            echo("API error: ${result.exceptionOrNull()?.message}", err = true)
            return
        }

        val response = result.getOrNull() ?: return
        val json = JsonParser.parseString(response).asJsonObject
        val candidates = json.getAsJsonArray("candidates")
        if (candidates == null || candidates.size() == 0) {
            echo("No response from Gemini", err = true)
            return
        }
        val text = candidates[0].asJsonObject
            .getAsJsonObject("content")
            .getAsJsonArray("parts")[0].asJsonObject
            .get("text")?.asString ?: ""
        echo(text)
    }
}
